File: py/wlearn/automl/_leaderboard.py
"""Leaderboard matching JS automl/leaderboard.js."""

import math

import numpy as np


class Leaderboard:
    def __init__(self):
        self._entries = []
        self._next_id = 0
        self._dirty = True

    def add(self, model_name, params, scores, fit_time_ms):
        """Add a candidate result.

        Args:
            model_name: str
            params: dict
            scores: np.ndarray of fold scores
            fit_time_ms: float

        Returns:
            entry dict
        """
        n = len(scores)
        mean_score = float(np.mean(scores))

        sum_sq = sum((scores[i] - mean_score) ** 2 for i in range(n))
        std_score = math.sqrt(sum_sq / n)

        entry = {
            'id': self._next_id,
            'modelName': model_name,
            'params': params,
            'scores': np.array(scores, dtype=np.float64),
            'meanScore': mean_score,
            'stdScore': std_score,
            'fitTimeMs': fit_time_ms,
            'rank': 0,
        }
        self._next_id += 1
        self._entries.append(entry)
        self._dirty = True
        return entry

    def ranked(self):
        """Return all entries sorted by meanScore descending with ranks."""
        if self._dirty:
            self._entries.sort(key=lambda e: -e['meanScore'])
            for i, entry in enumerate(self._entries):
                entry['rank'] = i + 1
            self._dirty = False
        return list(self._entries)

    def best(self):
        """Return the best entry or None."""
        if not self._entries:
            return None
        self.ranked()
        return self._entries[0]

    def top(self, k):
        """Return top k entries."""
        return self.ranked()[:k]

    def to_json(self):
        """Serialize to JSON-friendly list."""
        return [
            {
                'id': e['id'],
                'modelName': e['modelName'],
                'params': e['params'],
                'scores': list(e['scores']),
                'meanScore': e['meanScore'],
                'stdScore': e['stdScore'],
                'fitTimeMs': e['fitTimeMs'],
                'rank': e['rank'],
            }
            for e in self.ranked()
        ]

    @classmethod
    def from_json(cls, arr):
        """Deserialize from JSON array."""
        lb = cls()
        for e in arr:
            lb._entries.append({
                **e,
                'scores': np.array(e['scores'], dtype=np.float64),
            })
            if e['id'] >= lb._next_id:
                lb._next_id = e['id'] + 1
        lb._dirty = True
        return lb

    @property
    def length(self):
        return len(self._entries)

File: packages/automl/src/leaderboard.js
/**
 * Tracks and ranks candidate evaluation results.
 */
export class Leaderboard {
  #entries = []
  #nextId = 0
  #dirty = true

  /**
   * Add a candidate result.
   * @param {{ modelName: string, params: Record<string, unknown>, scores: Float64Array, fitTimeMs: number }} entry
   * @returns {object} the entry with id assigned
   */
  add({ modelName, params, scores, fitTimeMs }) {
    let sum = 0
    for (let i = 0; i < scores.length; i++) sum += scores[i]
    const meanScore = sum / scores.length

    let sumSq = 0
    for (let i = 0; i < scores.length; i++) {
      const d = scores[i] - meanScore
      sumSq += d * d
    }
    const stdScore = Math.sqrt(sumSq / scores.length)

    const entry = {
      id: this.#nextId++,
      modelName,
      params,
      scores,
      meanScore,
      stdScore,
      fitTimeMs,
      rank: 0,
    }
    this.#entries.push(entry)
    this.#dirty = true
    return entry
  }

  /**
   * Return all entries sorted by meanScore descending with ranks assigned.
   */
  ranked() {
    if (this.#dirty) {
      this.#entries.sort((a, b) => b.meanScore - a.meanScore)
      for (let i = 0; i < this.#entries.length; i++) {
        this.#entries[i].rank = i + 1
      }
      this.#dirty = false
    }
    return this.#entries.slice()
  }

  /**
   * Return the best entry (highest meanScore) or null.
   */
  best() {
    if (this.#entries.length === 0) return null
    this.ranked() // ensure sorted
    return this.#entries[0]
  }

  /**
   * Return top k entries.
   */
  top(k) {
    return this.ranked().slice(0, k)
  }

  /**
   * Serialize to JSON-friendly array.
   */
  toJSON() {
    return this.ranked().map(e => ({
      id: e.id,
      modelName: e.modelName,
      params: e.params,
      scores: [...e.scores],
      meanScore: e.meanScore,
      stdScore: e.stdScore,
      fitTimeMs: e.fitTimeMs,
      rank: e.rank,
    }))
  }

  /**
   * Deserialize from JSON array.
   */
  static fromJSON(arr) {
    const lb = new Leaderboard()
    for (const e of arr) {
      lb.#entries.push({
        ...e,
        scores: new Float64Array(e.scores),
      })
      if (e.id >= lb.#nextId) lb.#nextId = e.id + 1
    }
    lb.#dirty = true
    return lb
  }

  get length() {
    return this.#entries.length
  }
}
